src/Aspire.Hosting/Dashboard/proto/resource_service.proto

```diff
@@ -0,0 +1,170 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+syntax = "proto3";
+
+package aspire.v1;
+
+import "google/protobuf/timestamp.proto";
+
+////////////////////////////////////////////
+
+message ApplicationInformationRequest {
+}
+
+message ApplicationInformationResponse {
+    string application_name = 1;
+    string application_version = 2;
+}
+
+////////////////////////////////////////////
+
+message ResourceCommandRequest {
+    string command_type = 1;
+    ResourceId resource_id = 2;
+    optional string confirmation_message = 3;
+}
+
+enum ResourceCommandResponseKind {
+    none = 0;
+    succeeded = 1;
+    failed = 2;
+    cancelled = 3;
+}
+
+message ResourceCommandResponse {
+    ResourceCommandResponseKind kind = 1;
+    optional string error_message = 2;
+}
+
+////////////////////////////////////////////
+
+message ResourceType {
+    // Unique name for the resource type. Equivalent to ResourceId.resource_type
+    // If "display_name" is omitted, this value will be used in UIs.
+    string unique_name = 1;
+
+    // Display string for references to this type in UI. May be localized.
+    // If this value is omitted, UIs will show "unique_name" instead.
+    optional string display_name = 2;
+
+    // Any commands that may be executed against resources of this type, avoiding
+    // the need to copy the value to every Resource instance.
+    //
+    // Note that these commands must apply to matching resources at any time.
+    //
+    // If the set of commands changes over time, use the "commands" property
+    // of the Resource itself.
+    repeated ResourceCommandRequest commands = 3;
+}
+
+////////////////////////////////////////////
+
+message EnvironmentVariable {
+    string name = 1;
+    optional string value = 2;
+    bool is_from_spec = 3;
+}
+
+message Endpoint {
+    string name = 1;
+    optional string allocated_address = 2;
+    optional int32 allocated_port = 3;
+    optional string http_address = 4;
+}
+
+message StringArray {
+    repeated string values = 1;
+}
+
+message AdditionalData {
+    // Name of the data item, e.g. "id", "image", "path", ...
+    string name = 1;
+    // Optional namespace, e.g. "container", "executable", "project", ...
+    optional string namespace = 2;
+    // Optional display name, may be localized
+    optional string display_name = 3;
+    // A single value will be most common. Also support lists, to avoid escaping.
+    oneof kind {
+        string value = 4;
+        StringArray values = 5;
+    }
+}
+
+message ResourceId {
+    string uid = 1;
+    // TODO do we need resource_type to make unique names? if not, inline ResourceId type as string.
+    string resource_type = 2;
+}
+
+// Models the full state of an resource (container, executable, project, etc) at a particular point in time.
+message ResourceSnapshot {
```

Comment:
There will likely be some discrepancies between the values populated there locally and the ones available when running in the cloud. The `endpoints` field for example, won't be as open in the Cloud as what it is locally. I'm not sure yet of what we'll be able to provide there, but it's possible that we need to have some form of "nullable endpoint" concept here to say that an endpoint isn't available from outside of the cluster.

Reply:
Is there something similar to an endpoint you'd be interested in showing for such resources, or would the field just be omitted. Currently it's `repeated` where zero items is valid.

Another option is we move it to additional data.

Another option is we allow modelling hyperlinks for resources, similar to how we have commands. Maybe they're just a different kind of command (link vs. button).

Reply:
It would feel OK for me to have these endpoints being displayed prominently in the local dashboard, and being shown as "Internal"/"In-cluster" once it is deployed on Container Apps. I think that it makes sense for the developers to see such endpoints locally, and to have a more restricted experience once things are deployed publicly.
If we could have a way to specify this (rather than just setting the endpoint value to "Internal", which wouldn't be a valid URL) in the API it would be great however.

Reply:
@daniv-msft the current `Endpoint` message requires a name, but the addresses and port are optional. I think you could achieve what you wanted using the current design.